### DISCUSSION TITLE
How are you going to implement COM support in tB?

### DISCUSSION THREAD

**Discussion by YogiYang** [twinbasic/twinbasic#440] [General]
Hello,

I am just curious and want to know as to how are you going to add support for COM/ActiveX in tB.

Will it be like VB6?
-OR-
Will it be like VB.NET?

From what little that I understand about COM/ActiveX is that VB6 directly (read dynamically) interacts with the library and extracts necessary information and present it to the user. While VB.NET (and other similar dev tools) generates a COM Wrapper and the user has to use that wrapper to access the COM/ActiveX component.

**Comment by WaynePhillipsEA** [comment]
tB already supports COM and ActiveX libraries.  The support is native (i.e. no wrappers / interop or anything like that).

**Comment by bclothier** [comment]
Just to provide a little bit of context...

COM is not an API but an ABI & specifications which is why COM can be language-agnostic because it all come down to implementing your objects by arranging everything exactly the same. As long you follow the specifications, it will work basically anywhere and in any languages. This however require that you layout everything down to the last byte so that it's uniform and is not subject to different layouts due to compilers rearranging things around. Not all languages allow you that degree of control over the object's memory layout and for that reason, some languages require additional support from their compiler or such to guarantee the same layout.

However, .NET can be thought of as a separate environment in which it will track everything for garbage collection. That's why it's "managed" (whereas VB6, Python, JavaScript, C/C++, etc.  are all "unmanaged"). However, COM does have rules around how long you can keep a reference to object and when you must release the resources associated with that object. Those rules are deterministic --- as soon as you release the last reference to the object, it can be freed immediately. But in .NET, things are different -- instead of freeing resources on last release, it runs garbage collection in a separate thread, so the object may not be actually released until much later. Thus, that violates COM conventions in several ways and for that reason, we have wrappers which promise to adhere to the COM conventions and hiding the .NET's garbage collection behaviors from the COM clients. 

tB is unmanaged and uses reference-counting for garbage collection so it is able to comply with COM specifications without any kind of wrappers as is the case with .NET. Because backward compatibility is the priority number one, it will need to implement in the same way VBx has implemented its COM objects (and even beyond what is specified by COM -- VBx has additional behaviors that are not necessarily the part of the formal COM specifications).
